endpoints/web.py

```diff
@@ -984,3 +990,26 @@ def user_initialize():
         response = jsonify({"message": "Failed to initialize user: " + str(ex)})
         response.status_code = 400
         return response
+
+
+@web.route("/config", methods=["GET"])
+def config():
+    response = jsonify(
+        {
+            "config": frontend_visible_config(app.config),
+            "features": features.get_features(),
+            "oauth": get_oauth_config(),
+            "external_login": get_external_login_config(),
+            "registry_state": app.config.get("REGISTRY_STATE", "normal"),
+            "account_recovery_mode": app.config.get("ACCOUNT_RECOVERY_MODE", False),
+        }
+    )
+    return response
+
+
+@web.route("/csrf_token", methods=["GET"])
```

Comment:
it doesn't defeat the CSRF. It makes it easier for the frontend to access it. you can access the CSRF token today without any special permissions

```
curl https://quay.io | grep -i __token
```

This is needed for the new UI which runs as outside quay.

Reply:
If I'm understanding this correctly, the issue is that with the current ui, we can just include it in as a hidden form, session, etc. when serving the application.
But with the new ui being a separate domain, this becomes a more tricky. Like @dmage mentioned, it doesn't really make sense to allow the ui, which is from a different origin, to generate a csrf token.

Reply:
It kinda defeats the purpose of a csrf token in the first place.

Reply:
Instead, the new ui should be authenticating to Quay as a third party app to access its resources (oauth).

Reply:
Expanding on that, the token should be unique and associated to a session, so if the generator is exposed directly, it shouldn't matter since whatever token is generated won't be associated to another session.

Reply:
As I mentioned yesterday in the talks, this is a moot point if the new UI is hosted inside Quay's container as a separate nginx configuration. We already have specific nginx paths for specific endpoints (such as Quay's `v1` and Docker `v2` API). I don't see a reason why we wouldn't host the new UI with a specific worker behind a specific nginx path. That way the new UI shares the same domain and same backend workers as the current UI does so there are no CSRF shenanigans. If it's a matter of container size that's worrisome, then it's also not really an argument since the old UI will be going away at some point and new UI will be all that remains. I don't see a particular reason why the new UI would need to run in its own separate container.

Reply:
@ibazulic  Serving the UI out of the same container is not the way we want to move forward. There are many advantages of separating the UI from the backend. Sure, for now when we are in the transition from the old UI to the new UI we can serve the UI from the container to seamlessly switch between the two but once we phase out the old UI, we should serve the new UI outside the container. A few advantages:
1. Decoupling the UI and API makes it easier to develop as you are no longer tied to the same release cycle 
2. Since the UI is just a few assets, this can be hosted anywhere. For quay.io for example, this can be served out of a CDN which means we don't need a container for the UI
3. Since it can be hosted anywhere, integration with the OCP console for example would become much easier as we can just send assets to them and all the communication will happen over the API
4. It makes us think in terms of API rather than UI. Meaning, whenever we develop a new feature that we want to expose in the UI, the API will be the only way to expose that data

Reply:
There are two kind of authentication mechanisms: token-based and session-based.

CSRF tokens are only relevant for session-based authentication: the website should perform actions based on the session only if the CSRF token is valid. Otherwise a malicious site can use a simple form `<from method="post" action="https://quay.io/destory/everything">` and trick the user to submit the form.

For token-based (i.e. OAuth) there is no need in CSRF tokens.

Another way to be protected from CSRF is to use `Lax` or `Strict` cookies.

> Expanding on that, the token should be unique and associated to a session, so if the generator is exposed directly, it shouldn't matter since whatever token is generated won't be associated to another session

It's not about another session, it's about your session. If a malicious site can obtain a CSRF token via Ajax, the CSRF protection is devastated: the malicious site, once you visit it, can do anything on Quay on your behalf via your browser.

> I'm trying to understand how it's different from our current situation.

It's not different. It's just not the accepted way to do things.

If you have a mechanism for this CSRF endpoint, you can use it for other API endpoints and get rid of the CSRF endpoint. You won't get a better protection because of an extra API call. You'll get security through obscurity/complexity. 

> A 3rd domain can just as easily get the CSRF token by doing a `get` on any quay page. CSRF token protects us because we don't allow cross origin requests to send a `Cookie` in the request headers when the origin is set to `*` so, even if they have access to the CSRF token via the `get` it's pretty much unusable because they are not in the correct domain. Exposing the CSRF via API doesn't bypass this security measure

The malicious website can obtain a CSRF token via Ajax and inject it into a traditional form. Quay's session cookie is Lax, so the attacker can succeed only with ~6% of browsers that are very old and don't support SameSite=Lax cookies (https://caniuse.com/mdn-http_headers_set-cookie_samesite_lax).

Reply:
CSRF tokens != authentication tokens. A new token doesn't give you a new session.

Reply:
@syed 
> 
> 1. Decoupling the UI and API makes it easier to develop as you are no longer tied to the same release cycle

That means creating new functions that will then call specific API functions, doesn't it? Seems to me this only introduces complexity. Current model makes sure that all UI functions use the same API requests as the end user would do via `curl`. 

> 2. Since the UI is just a few assets, this can be hosted anywhere. For quay.io for example, this can be served out of a CDN which means we don't need a container for the UI

Assets are not really important here. You still need a web server that will serve requests for this UI. The UI will be run in something, presumably Flask but even if it's something else, you still need a container. Sure, it will be a smaller container but it will still exist. 

> 3. Since it can be hosted anywhere, integration with the OCP console for example would become much easier as we can just send assets to them and all the communication will happen over the API

But we have a lot of customers that aren't really running Quay inside OpenShift. So they will need that console container. 

> 4. It makes us think in terms of API rather than UI. Meaning, whenever we develop a new feature that we want to expose in the UI, the API will be the only way to expose that data

Well, Quay was created from the grounds up with API in mind, onto which other things were added. Which is why the UI today uses the same API endpoints and same API functions as end user do via `curl`.

Reply:
@dmage 
> It's not about another session, it's about your session. If a malicious site can obtain a CSRF token via Ajax, the CSRF protection is devastated: the malicious site, once you visit it, can do anything on Quay on your behalf via your browser.

In your example, the attacker cannot obtain the CSRF token. Let's assume the attacker does an ajax call to `/csrf_token` endpoint. The attacker will get a CSRF token which is different from the CSRF token that belongs to your session. They cannot override that. Sure they can inject the CSRF token into a form and send that request to quay but that request will never get accepted because that CSRF token will not match the CSRF token that Quay has associated with the original user session. 

Again, I don't even need the `/csrf_token` endpoint to get the CSRF token. If what you are saying is true,  then quay is vulnerable today because any page you request from quay.io will have a CSRF token embedded in it. If I was an attacker I would only have to do one extra step of greping the token out of the HTML response.

Reply:
@ibazulic 

> Current model makes sure that all UI functions use the same API requests as the end user would do via curl.

> Well, Quay was created from the grounds up with API in mind, onto which other things were added. Which is why the UI today uses the same API endpoints and same API functions as end user do via curl.


No, the current model injects a bunch of hidden variables in the page that gets rendered on the UI. These variables are not available via an API. This is because we haven't decoupled the UI completely from the backend. Decoupling means the only way the frontend and backend communicate is via the API which is not happening today.

> You still need a web server that will serve requests for this UI. The UI will be run in something, presumably Flask but even if it's something else, you still need a container. Sure, it will be a smaller container but it will still exist.

React builds into a single JS file. So really we just need something to serve static assets. For single page applications like react the best practice these days is to serve them from object storage directly avoiding the need for a web server. But to your point about people who don't run Quay in openshift. The frontend bundle can be included in the Quay container

Reply:
@syed 

> In your example, the attacker cannot obtain the CSRF token.

Oh, so this endpoint returns just a random string that is not a valid CSRF token (SameSite=Lax should restrict it to Quay's domain). No valid token — no attack. That's not my point though.

Can you elaborate on why you need this endpoint? You shouldn't use a cross-site request for protection against cross-site request forgery.

Reply:
@dmage it returns a valid CSRF token for a session only if you pass a valid session cookie. Otherwise, it will generate a random CSRF token. The API endpoint is needed for the new UI. It is based on react and is a SPA. Currently we inject the CSRF token on the server side while rendering the HTML. But that's the old way. With new web apps they purely communicate to the backend over an API so we can't render the CSRF token in the HTML. 

Exposing the CSRF token generator is not a security risk at all. As I said in my previous post we do it right now here in `base.html` which is rendered for all HTML responses from Quay https://github.com/quay/quay/blob/master/templates/base.html#L47

Reply:
How will the session initially get passed client side? Couldn't you just pass the token whenever that happens? Not sure I understand why that endpoint is needed.